Error: "Module '"[...]/@types/react/index"' can only be default-imported using the 'allowSyntheticDefaultImports' flag"

<!--
  Thanks for filing an issue on Apollo Client!

  Please make sure that you include the following information to ensure that your issue is actionable.

  If you don't follow the template, your issue may end up being closed without anyone looking at it carefully, because it is not actionable for us without the information in this template.

  **PLEASE NOTE:** Feature requests and non-bug related discussions are no longer managed in this repo. Feature requests should be opened in https://github.com/apollographql/apollo-feature-requests.
-->

**Intended outcome:**
<!--
What you were trying to accomplish when the bug occurred, and as much code as possible related to the source of the problem.
-->

I should be able to import `@apollo/client` without needing `allowSyntheticDefaultImports` or `esModuleInterop` set in my tsconfig.

**Actual outcome:**
<!--
A description of what actually happened, including a screenshot or copy-paste of any related error messages, logs, or other output that might be related. Places to look for information include your browser console, server console, and network logs. Please avoid non-specific phrases like “didn’t work” or “broke”.
-->

When I import `@apollo/client` without the `allowSyntheticDefaultImports` and `esModuleImports` flags set in my tsconfig, I get a series of errors like this:

```
../../node_modules/@apollo/client/react/context/ApolloConsumer.d.ts:1:8 - error TS1259: Module '"/path/to/project/node_modules/@types/react/index"' can only be default-imported using the 'allowSyntheticDefaultImports' flag

1 import React from 'react';
         ~~~~~

  ../../node_modules/@types/react/index.d.ts:65:1
    65 export = React;
       ~~~~~~~~~~~~~~~
    This module is declared with using 'export =', and can only be used with a default import when using the 'allowSyntheticDefaultImports' flag.
```

Similar errors are produced for both `react` and `zen-observable`.

**How to reproduce the issue:**
<!--
If possible, please create a reproduction using https://github.com/apollographql/react-apollo-error-template and link to it here. If you prefer an in-browser way to create reproduction, try: https://codesandbox.io/s/github/apollographql/react-apollo-error-template

Instructions for how the issue can be reproduced by a maintainer or contributor. Be as specific as possible, and only mention what is necessary to reproduce the bug. If possible, try to isolate the exact circumstances in which the bug occurs and avoid speculation over what the cause might be.
-->

To Reproduce:
1. Install `@apollo/client`, `graphql`, and `@types/react` in a typescript-based repo
2. Import from `@apollo/client` (I reproduced with `ApolloConsumer`, but it should also be reproducible with imports)
3. Compile the file that imports from `@apollo/client` using the typescript compiler

Reproduction Example:
1. Clone https://github.com/devrelm/apollo-client-synthetic-defaults-example
2. Run `yarn run tsc`

**Versions**
<!--
Run the following command in your project directory, and paste its (automatically copied to clipboard) results here:

`npx envinfo@latest --preset apollo --clipboard`
-->
```
  System:
    OS: macOS 10.15.7
  Binaries:
    Node: 12.18.3 - /usr/local/bin/node
    Yarn: 1.22.4 - /usr/local/bin/yarn
    npm: 6.14.6 - /usr/local/bin/npm
  Browsers:
    Chrome: 88.0.4324.182
    Firefox: 85.0.1
    Safari: 14.0.2
  npmPackages:
    @apollo/client: ^3.3.11 => 3.3.11
```

---

**Other Notes**

It looks like this has already been brought up a few times but hasn't been totally fixed yet:

* #7004 PR: "Fix zen-observable import" - closed as the author ended up adding `allowSyntheticDefaultImports` to fix the error
* #7005 Issue: "TypeScript error when using client without react" - Still open, but I don't think this is necessarily a duplicate issue, as that one is about *not* using react and so has additional possible solutions. This present issue cannot be solved the same way since I actually *do* want to use react.
* 78bc4bcc tried to fix the issue, but it was reverted by @benjamn in 113475b1 with the message This reverts commit [78bc4bcc], which caused more problems than it solved.

I'll try to get a PR in to fix the issue, but it would be nice to know how the fix "caused more problems than it solved" so that I could avoid any of the same pitfalls.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -26,6 +26,9 @@ TBD
 - Allow `merge: true` field policy to merge `Reference` objects with non-normalized objects, and vice-versa. <br/>
   [@benjamn](https://github.com/benjamn) in [#7778](https://github.com/apollographql/apollo-client/pull/7778)
 
+- Prefer `import * as namepace ...` for imports whose types are re-exported (and thus may appear in `.d.ts` files). <br/>
+  [@devrelm](https://github.com/devrelm) in [#7742](https://github.com/apollographql/apollo-client/pull/7742)
+
 ### Documentation
 TBD
 

--- a/src/link/persisted-queries/__tests__/react.tsx
+++ b/src/link/persisted-queries/__tests__/react.tsx
@@ -1,5 +1,5 @@
-import React from 'react';
-import ReactDOM from 'react-dom/server';
+import * as React from 'react';
+import * as ReactDOM from 'react-dom/server';
 import gql from 'graphql-tag';
 import { print } from 'graphql';
 import { sha256 } from 'crypto-hash';

--- a/src/react/components/Mutation.tsx
+++ b/src/react/components/Mutation.tsx
@@ -1,4 +1,4 @@
-import PropTypes from 'prop-types';
+import * as PropTypes from 'prop-types';
 
 import { OperationVariables } from '../../core';
 import { MutationComponentOptions } from './types';

--- a/src/react/components/Query.tsx
+++ b/src/react/components/Query.tsx
@@ -1,4 +1,4 @@
-import PropTypes from 'prop-types';
+import * as PropTypes from 'prop-types';
 
 import { OperationVariables } from '../../core';
 import { QueryComponentOptions } from './types';

--- a/src/react/components/Subscription.tsx
+++ b/src/react/components/Subscription.tsx
@@ -1,4 +1,4 @@
-import PropTypes from 'prop-types';
+import * as PropTypes from 'prop-types';
 
 import { OperationVariables } from '../../core';
 import { SubscriptionComponentOptions } from './types';

--- a/src/react/context/ApolloConsumer.tsx
+++ b/src/react/context/ApolloConsumer.tsx
@@ -1,4 +1,4 @@
-import React from 'react';
+import * as React from 'react';
 import { invariant } from 'ts-invariant';
 
 import { ApolloClient } from '../../core';

--- a/src/react/context/ApolloContext.ts
+++ b/src/react/context/ApolloContext.ts
@@ -1,4 +1,4 @@
-import React from 'react';
+import * as React from 'react';
 import { ApolloClient } from '../../core';
 import { canUseWeakMap } from '../../utilities';
 

--- a/src/react/context/ApolloProvider.tsx
+++ b/src/react/context/ApolloProvider.tsx
@@ -1,4 +1,4 @@
-import React from 'react';
+import * as React from 'react';
 import { invariant } from 'ts-invariant';
 
 import { ApolloClient } from '../../core';

--- a/src/react/hoc/hoc-utils.tsx
+++ b/src/react/hoc/hoc-utils.tsx
@@ -1,4 +1,4 @@
-import React from 'react';
+import * as React from 'react';
 import { invariant } from 'ts-invariant';
 import { OperationVariables } from '../../core';
 import { IDocumentDefinition } from '../parser';

--- a/src/react/hoc/mutation-hoc.tsx
+++ b/src/react/hoc/mutation-hoc.tsx
@@ -1,4 +1,4 @@
-import React from 'react';
+import * as React from 'react';
 import { DocumentNode } from 'graphql';
 import hoistNonReactStatics from 'hoist-non-react-statics';
 

--- a/src/react/hoc/query-hoc.tsx
+++ b/src/react/hoc/query-hoc.tsx
@@ -1,4 +1,4 @@
-import React from 'react';
+import * as React from 'react';
 import { DocumentNode } from 'graphql';
 import hoistNonReactStatics from 'hoist-non-react-statics';
 

--- a/src/react/hoc/subscription-hoc.tsx
+++ b/src/react/hoc/subscription-hoc.tsx
@@ -1,4 +1,4 @@
-import React from 'react';
+import * as React from 'react';
 import { DocumentNode } from 'graphql';
 import hoistNonReactStatics from 'hoist-non-react-statics';
 

--- a/src/react/hoc/withApollo.tsx
+++ b/src/react/hoc/withApollo.tsx
@@ -1,4 +1,4 @@
-import React from 'react';
+import * as React from 'react';
 import hoistNonReactStatics from 'hoist-non-react-statics';
 import { invariant } from 'ts-invariant';
 

--- a/src/react/hooks/useApolloClient.ts
+++ b/src/react/hooks/useApolloClient.ts
@@ -1,4 +1,4 @@
-import React from 'react';
+import * as React from 'react';
 import { invariant } from 'ts-invariant';
 
 import { ApolloClient } from '../../core';

--- a/src/react/ssr/getDataFromTree.ts
+++ b/src/react/ssr/getDataFromTree.ts
@@ -1,4 +1,4 @@
-import React from 'react';
+import * as React from 'react';
 import { getApolloContext } from '../context';
 import { RenderPromises } from './RenderPromises';
 

--- a/src/utilities/testing/mocking/MockedProvider.tsx
+++ b/src/utilities/testing/mocking/MockedProvider.tsx
@@ -1,4 +1,4 @@
-import React from 'react';
+import * as React from 'react';
 
 import { ApolloClient, DefaultOptions } from '../../../core';
 import { InMemoryCache as Cache } from '../../../cache';

